File: src/CSharp/src/Durable7.FingerTree/BilateralAncestralDeque.cs
using System.Collections;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;

namespace Durable7.FingerTree;

/// <summary>
/// Represents an immutable deque as two oppositely oriented intervals in an append-only ancestry
/// arena.
/// </summary>
/// <typeparam name="T">The element type.</typeparam>
/// <remarks>
/// <para>
/// Each handle stores a left interval (l-a, l-t] and a right interval (r-a, r-t]. Its logical
/// value is reverse((l-a, l-t]) followed by (r-a, r-t]. A front insertion appends a leaf to the
/// left tail; a back insertion appends one to the right tail. Reversal only exchanges the two
/// intervals. Every contiguous slice intersects each interval at most once, so it is again a
/// bilateral handle rather than a rebuilt tree.
/// </para>
/// <para>
/// Let U be arena leaf-add time and Q be arena level-ancestor query time. End insertion costs U;
/// removal costs O(1) on its own nonempty side and at most one Q when it crosses the center;
/// indexing costs at most one Q; slicing and splitting each cost at most two Q.
/// Endpoint reads, count, clear, and reverse are O(1). An Alstrup--Holm incremental-level-ancestor
/// backend makes all of those operations O(1) worst case with linear arena space. The shipped
/// <see cref="MyersIncrementalAncestorArena{T}"/> instead gives O(1)-amortized insertion and
/// O(log M) ancestor queries after M historical insertions.
/// </para>
/// <para>
/// The restricted algebra intentionally excludes arbitrary concatenation, point replacement, and
/// middle editing. Arena space is charged to all successful historical end insertions, not merely
/// to the values visible in one handle. Enumeration is Theta(Count) time and may retain
/// O(Count) temporary values. Handles are immutable; progress guarantees come from the arena.
/// </para>
/// </remarks>
[DebuggerDisplay("Count = {Count}, Left = {_left.Count}, Right = {_right.Count}")]
public sealed class BilateralAncestralDeque<T> : IReadOnlyList<T>
{
    private readonly IIncrementalAncestorArena<T> _arena;
    private readonly Segment _left;
    private readonly Segment _right;

    private BilateralAncestralDeque(
        IIncrementalAncestorArena<T> arena,
        Segment left,
        Segment right,
        int count)
    {
        _arena = arena;
        _left = left;
        _right = right;
        Count = count;
    }

    /// <summary>Gets the number of visible values.</summary>
    public int Count { get; }

    /// <summary>Gets whether this handle denotes an empty deque.</summary>
    public bool IsEmpty => Count == 0;

    /// <summary>Gets the first visible value in O(1) arena operations.</summary>
    /// <exception cref="InvalidOperationException">The deque is empty.</exception>
    public T First
    {
        get
        {
            EnsureNotEmpty();
            return _arena.GetValue(_left.Count != 0 ? _left.Tail : _right.Base);
        }
    }

    /// <summary>Gets the last visible value in O(1) arena operations.</summary>
    /// <exception cref="InvalidOperationException">The deque is empty.</exception>
    public T Last
    {
        get
        {
            EnsureNotEmpty();
            return _arena.GetValue(_right.Count != 0 ? _right.Tail : _left.Base);
        }
    }

    /// <summary>Gets a visible value with at most one level-ancestor query.</summary>
    /// <param name="index">The zero-based visible index.</param>
    /// <returns>The value at <paramref name="index"/>.</returns>
    /// <exception cref="ArgumentOutOfRangeException"><paramref name="index"/> is outside the deque.</exception>
    public T this[int index]
    {
        get
        {
            if ((uint)index >= (uint)Count)
                throw IndexError(index);

            int node;
            if (index < _left.Count)
            {
                if (index == 0)
                    node = _left.Tail;
                else if (index == _left.Count - 1)
                    node = _left.Base;
                else
                    node = _arena.AncestorAtDepth(_left.Tail, _arena.GetDepth(_left.Tail) - index);
            }
            else
            {
                var rightIndex = index - _left.Count;
                if (rightIndex == 0)
                    node = _right.Base;
                else if (rightIndex == _right.Count - 1)
                    node = _right.Tail;
                else
                    node = _arena.AncestorAtDepth(
                        _right.Tail,
                        checked(_arena.GetDepth(_right.Anchor) + rightIndex + 1));
            }

            return _arena.GetValue(node);
        }
    }

    /// <summary>Creates an empty deque owned by an explicit incremental-ancestor arena.</summary>
    /// <param name="arena">The manager that retains and navigates every inserted node.</param>
    /// <returns>An empty bilateral handle.</returns>
    /// <exception cref="ArgumentNullException"><paramref name="arena"/> is null.</exception>
    /// <exception cref="ArgumentException">The arena's bottom node is not at depth -1.</exception>
    public static BilateralAncestralDeque<T> Create(IIncrementalAncestorArena<T> arena)
    {
        ArgumentNullException.ThrowIfNull(arena);
        var bottom = arena.Bottom;
        if (arena.GetDepth(bottom) != -1)
            throw new ArgumentException("The arena bottom node must have depth -1.", nameof(arena));
        var empty = Segment.EmptyAt(bottom);
        return new BilateralAncestralDeque<T>(arena, empty, empty, 0);
    }

    /// <summary>Creates an empty deque backed by the shipped Myers jump-link arena.</summary>
    /// <returns>An empty deque with a fresh, independently collectible arena.</returns>
    public static BilateralAncestralDeque<T> CreateMyers() => Create(new MyersIncrementalAncestorArena<T>());

    /// <summary>Creates a Myers-backed deque from values in enumeration order.</summary>
    /// <param name="values">The values to append.</param>
    /// <returns>A deque containing the values.</returns>
    /// <exception cref="ArgumentNullException"><paramref name="values"/> is null.</exception>
    public static BilateralAncestralDeque<T> CreateRange(IEnumerable<T> values)
    {
        ArgumentNullException.ThrowIfNull(values);
        var result = CreateMyers();
        foreach (var value in values)
            result = result.AddLast(value);
        return result;
    }

    /// <summary>Returns a deque with one value added at the front.</summary>
    /// <param name="value">The value to prepend.</param>
    /// <returns>The extended persistent branch.</returns>
    /// <exception cref="OverflowException">The visible count or ancestry depth cannot grow further.</exception>
    public BilateralAncestralDeque<T> AddFirst(T value)
    {
        EnsureRoom();
        var left = AddToTail(_left, value);
        return new BilateralAncestralDeque<T>(_arena, left, _right, Count + 1);
    }

    /// <summary>Returns a deque with one value added at the back.</summary>
    /// <param name="value">The value to append.</param>
    /// <returns>The extended persistent branch.</returns>
    /// <exception cref="OverflowException">The visible count or ancestry depth cannot grow further.</exception>
    public BilateralAncestralDeque<T> AddLast(T value)
    {
        EnsureRoom();
        var right = AddToTail(_right, value);
        return new BilateralAncestralDeque<T>(_arena, _left, right, Count + 1);
    }

    /// <summary>
    /// Returns the deque without its first value, using at most one level-ancestor query.
    /// </summary>
    /// <returns>The remaining deque.</returns>
    /// <exception cref="InvalidOperationException">The deque is empty.</exception>
    public BilateralAncestralDeque<T> RemoveFirst()
    {
        EnsureNotEmpty();
        if (Count == 1)
            return EmptyHandle();
        if (_left.Count != 0)
        {
            var left = RemoveTail(_left);
            return new BilateralAncestralDeque<T>(_arena, left, _right, Count - 1);
        }

        var right = RemoveBase(_right);
        return new BilateralAncestralDeque<T>(_arena, _left, right, Count - 1);
    }

    /// <summary>
    /// Returns the deque without its last value, using at most one level-ancestor query.
    /// </summary>
    /// <returns>The remaining deque.</returns>
    /// <exception cref="InvalidOperationException">The deque is empty.</exception>
    public BilateralAncestralDeque<T> RemoveLast()
    {
        EnsureNotEmpty();
        if (Count == 1)
            return EmptyHandle();
        if (_right.Count != 0)
        {
            var right = RemoveTail(_right);
            return new BilateralAncestralDeque<T>(_arena, _left, right, Count - 1);
        }

        var left = RemoveBase(_left);
        return new BilateralAncestralDeque<T>(_arena, left, _right, Count - 1);
    }

    /// <summary>Attempts to remove and return the first value.</summary>
    /// <param name="value">The removed value on success; otherwise default.</param>
    /// <param name="result">The remaining deque on success; otherwise this deque.</param>
    /// <returns><see langword="true"/> when a value was removed.</returns>
    public bool TryRemoveFirst(
        [MaybeNullWhen(false)] out T value,
        out BilateralAncestralDeque<T> result)
    {
        if (IsEmpty)
        {
            value = default;
            result = this;
            return false;
        }

        value = First;
        result = RemoveFirst();
        return true;
    }

    /// <summary>Attempts to remove and return the last value.</summary>
    /// <param name="value">The removed value on success; otherwise default.</param>
    /// <param name="result">The remaining deque on success; otherwise this deque.</param>
    /// <returns><see langword="true"/> when a value was removed.</returns>
    public bool TryRemoveLast(
        [MaybeNullWhen(false)] out T value,
        out BilateralAncestralDeque<T> result)
    {
        if (IsEmpty)
        {
            value = default;
            result = this;
            return false;
        }

        value = Last;
        result = RemoveLast();
        return true;
    }

    /// <summary>Returns the first <paramref name="count"/> values.</summary>
    /// <param name="count">The prefix length.</param>
    /// <returns>A persistent bilateral slice.</returns>
    /// <exception cref="ArgumentOutOfRangeException">
    /// <paramref name="count"/> is outside zero through <see cref="Count"/>.
    /// </exception>
    public BilateralAncestralDeque<T> Take(int count)
    {
        ValidateBoundary(count, nameof(count));
        return count == Count ? this : Slice(0, count);
    }

    /// <summary>Returns the deque after omitting its first <paramref name="count"/> values.</summary>
    /// <param name="count">The number of values to omit.</param>
    /// <returns>A persistent bilateral slice.</returns>
    /// <exception cref="ArgumentOutOfRangeException">
    /// <paramref name="count"/> is outside zero through <see cref="Count"/>.
    /// </exception>
    public BilateralAncestralDeque<T> Drop(int count)
    {
        ValidateBoundary(count, nameof(count));
        return count == 0 ? this : Slice(count, Count - count);
    }

    /// <summary>Returns one contiguous slice using at most two level-ancestor queries.</summary>
    /// <param name="index">The zero-based first visible index.</param>
    /// <param name="count">The number of visible values in the result.</param>
    /// <returns>The selected bilateral ancestry intervals.</returns>
    /// <exception cref="ArgumentOutOfRangeException">
    /// The requested index/count pair is outside this deque.
    /// </exception>
    public BilateralAncestralDeque<T> Slice(int index, int count)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index);
        ArgumentOutOfRangeException.ThrowIfNegative(count);
        if (index > Count)
            throw new ArgumentOutOfRangeException(nameof(index));
        if (index > Count - count)
            throw new ArgumentOutOfRangeException(nameof(count));
        if (index == 0 && count == Count)
            return this;
        if (count == 0)
            return EmptyHandle();

        var end = index + count;
        var leftStart = Math.Min(index, _left.Count);
        var leftEnd = Math.Min(end, _left.Count);
        var rightStart = Math.Max(0, index - _left.Count);
        var rightEnd = Math.Max(0, end - _left.Count);

        var left = SliceLeft(_left, leftStart, leftEnd - leftStart);
        var right = SliceRight(_right, rightStart, rightEnd - rightStart);
        return new BilateralAncestralDeque<T>(_arena, left, right, count);
    }

    /// <summary>Splits this deque at a zero-based boundary.</summary>
    /// <param name="index">The number of values placed in the left result.</param>
    /// <returns>The prefix and suffix; together they enumerate the original value.</returns>
    /// <exception cref="ArgumentOutOfRangeException">
    /// <paramref name="index"/> is outside zero through <see cref="Count"/>.
    /// </exception>
    public (BilateralAncestralDeque<T> Left, BilateralAncestralDeque<T> Right) SplitAt(int index)
    {
        ValidateBoundary(index, nameof(index));
        return (Take(index), Drop(index));
    }

    /// <summary>Returns the logical reversal by exchanging the two oriented intervals.</summary>
    /// <returns>The reversed persistent handle.</returns>
    public BilateralAncestralDeque<T> Reverse()
    {
        if (Count <= 1)
            return this;
        return new BilateralAncestralDeque<T>(_arena, _right, _left, Count);
    }

    /// <summary>Returns an empty handle in the same arena.</summary>
    /// <returns>An independently appendable empty deque.</returns>
    public BilateralAncestralDeque<T> Clear()
    {
        if (IsEmpty)
            return this;
        return EmptyHandle();
    }

    /// <summary>Copies the values to a new array in logical order.</summary>
    /// <returns>A new array containing this deque's values.</returns>
    public T[] ToArray()
    {
        if (IsEmpty)
            return [];

        var values = new T[Count];
        var node = _left.Tail;
        for (var index = 0; index < _left.Count; index++)
        {
            values[index] = _arena.GetValue(node);
            if (index + 1 != _left.Count)
                node = _arena.GetParent(node);
        }

        node = _right.Tail;
        for (var index = Count - 1; index >= _left.Count; index--)
        {
            values[index] = _arena.GetValue(node);
            if (index != _left.Count)
                node = _arena.GetParent(node);
        }

        return values;
    }

    /// <summary>Returns a stable front-to-back enumerator over this immutable handle.</summary>
    /// <returns>An enumerator.</returns>
    /// <remarks>
    /// Enumeration takes Theta(Count) time. The forward-oriented right interval is reversed through
    /// a temporary array, so the enumerator can retain O(Count) values in the worst case.
    /// </remarks>
    public IEnumerator<T> GetEnumerator() => Enumerate().GetEnumerator();

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    /// <summary>Validates cached interval endpoints and counts. Test-only; O(1) queries.</summary>
    internal void ValidateInvariants()
    {
        if (Count != checked(_left.Count + _right.Count))
            throw new InvalidOperationException("The total count disagrees with the interval counts.");
        ValidateSegment(_left);
        ValidateSegment(_right);
    }

    private IEnumerable<T> Enumerate()
    {
        var node = _left.Tail;
        for (var index = 0; index < _left.Count; index++)
        {
            yield return _arena.GetValue(node);
            if (index + 1 != _left.Count)
                node = _arena.GetParent(node);
        }

        if (_right.Count == 0)
            yield break;

        var rightValues = new T[_right.Count];
        node = _right.Tail;
        for (var index = _right.Count - 1; index >= 0; index--)
        {
            rightValues[index] = _arena.GetValue(node);
            if (index != 0)
                node = _arena.GetParent(node);
        }

        foreach (var value in rightValues)
            yield return value;
    }

    private Segment AddToTail(Segment segment, T value)
    {
        var tail = _arena.AddLeaf(segment.Tail, value);
        var @base = segment.Count == 0 ? tail : segment.Base;
        return new Segment(segment.Anchor, @base, tail, segment.Count + 1);
    }

    private Segment RemoveTail(Segment segment)
    {
        Debug.Assert(segment.Count != 0);
        var tail = _arena.GetParent(segment.Tail);
        if (segment.Count == 1)
            return Segment.EmptyAt(tail);
        return new Segment(segment.Anchor, segment.Base, tail, segment.Count - 1);
    }

    private Segment RemoveBase(Segment segment)
    {
        Debug.Assert(segment.Count != 0);
        if (segment.Count == 1)
            return Segment.EmptyAt(segment.Tail);

        var newAnchor = segment.Base;
        var newBase = segment.Count == 2
            ? segment.Tail
            : _arena.AncestorAtDepth(
                segment.Tail,
                checked(_arena.GetDepth(segment.Tail) - segment.Count + 2));
        return new Segment(newAnchor, newBase, segment.Tail, segment.Count - 1);
    }

    private Segment SliceLeft(Segment segment, int start, int count)
    {
        Debug.Assert(start >= 0 && count >= 0 && start <= segment.Count - count);
        if (count == 0)
            return Segment.EmptyAt(_arena.Bottom);
        if (start == 0 && count == segment.Count)
            return segment;

        var tailDepth = _arena.GetDepth(segment.Tail);
        var baseDepth = tailDepth - segment.Count + 1;
        var slicedTailDepth = tailDepth - start;
        var slicedBaseDepth = slicedTailDepth - count + 1;

        var tail = slicedTailDepth == tailDepth
            ? segment.Tail
            : _arena.AncestorAtDepth(segment.Tail, slicedTailDepth);
        int @base;
        if (slicedBaseDepth == slicedTailDepth)
            @base = tail;
        else if (slicedBaseDepth == baseDepth)
            @base = segment.Base;
        else
            @base = _arena.AncestorAtDepth(segment.Tail, slicedBaseDepth);

        var anchor = @base == segment.Base ? segment.Anchor : _arena.GetParent(@base);
        return new Segment(anchor, @base, tail, count);
    }

    private Segment SliceRight(Segment segment, int start, int count)
    {
        Debug.Assert(start >= 0 && count >= 0 && start <= segment.Count - count);
        if (count == 0)
            return Segment.EmptyAt(_arena.Bottom);
        if (start == 0 && count == segment.Count)
            return segment;

        var anchorDepth = _arena.GetDepth(segment.Anchor);
        var slicedBaseDepth = checked(anchorDepth + start + 1);
        var slicedTailDepth = checked(slicedBaseDepth + count - 1);
        var end = start + count;

        var @base = start == 0
            ? segment.Base
            : _arena.AncestorAtDepth(segment.Tail, slicedBaseDepth);
        int tail;
        if (slicedTailDepth == slicedBaseDepth)
            tail = @base;
        else if (end == segment.Count)
            tail = segment.Tail;
        else
            tail = _arena.AncestorAtDepth(segment.Tail, slicedTailDepth);

        var anchor = start == 0 ? segment.Anchor : _arena.GetParent(@base);
        return new Segment(anchor, @base, tail, count);
    }

    private void ValidateSegment(Segment segment)
    {
        if (segment.Count == 0)
        {
            if (segment.Anchor != segment.Base || segment.Base != segment.Tail)
                throw new InvalidOperationException("An empty interval does not have one shared endpoint.");
            return;
        }

        var anchorDepth = _arena.GetDepth(segment.Anchor);
        var tailDepth = _arena.GetDepth(segment.Tail);
        if (tailDepth - anchorDepth != segment.Count)
            throw new InvalidOperationException("An interval count disagrees with its endpoint depths.");
        if (_arena.GetParent(segment.Base) != segment.Anchor)
            throw new InvalidOperationException("The cached base is not the first node after the anchor.");
        if (_arena.AncestorAtDepth(segment.Tail, anchorDepth) != segment.Anchor)
            throw new InvalidOperationException("The interval anchor is not an ancestor of its tail.");
        if (_arena.AncestorAtDepth(segment.Tail, anchorDepth + 1) != segment.Base)
            throw new InvalidOperationException("The cached base is not on the tail's ancestry path.");
    }

    private void ValidateBoundary(int index, string parameterName)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(index, parameterName);
        if (index > Count)
            throw new ArgumentOutOfRangeException(parameterName);
    }

    private void EnsureRoom()
    {
        if (Count == int.MaxValue)
            throw new OverflowException("The deque count exceeds Int32.MaxValue.");
    }

    private BilateralAncestralDeque<T> EmptyHandle()
    {
        var empty = Segment.EmptyAt(_arena.Bottom);
        return new BilateralAncestralDeque<T>(_arena, empty, empty, 0);
    }

    private void EnsureNotEmpty()
    {
        if (IsEmpty)
            throw new InvalidOperationException("The bilateral ancestral deque is empty.");
    }

    private static ArgumentOutOfRangeException IndexError(int index) =>
        new(nameof(index), index, "Index is outside the valid range of the deque.");

    private readonly record struct Segment(int Anchor, int Base, int Tail, int Count)
    {
        internal static Segment EmptyAt(int node) => new(node, node, node, 0);
    }
}

File: src/OCaml/lib/finger_tree/bilateral_ancestral_deque.ml
(** Bilateral ancestral deque: a persistent deque held as two oppositely oriented intervals of an
    append-only ancestry arena.

    A handle stores two interval descriptors over arena node handles, so it is immutable and
    constant-sized even though the arena underneath it grows. Every operation is the interval
    arithmetic of the C# baseline transcribed directly; the only structure this module owns is the
    pair of descriptors. *)

type segment = { anchor : int; base : int; tail : int; size : int }
(** One oriented ancestry interval running from the exclusive [anchor] down to the inclusive [tail].
    [base] is redundant but load-bearing: it caches the first physical node after [anchor], which
    keeps both public endpoint reads query-free even when the opposite arm is empty. For the left
    arm the logical first value sits at [tail] and the logical last at [base]; for the right arm
    those roles are exchanged. *)

type 'value t = {
  arena : 'value Incremental_ancestor.t;
  left : segment;
  right : segment;
  total : int;
}

type statistics = { count : int; left_count : int; right_count : int }

let ( let* ) = Result.bind

(** The empty interval anchored at one node. *)
let empty_at node = { anchor = node; base = node; tail = node; size = 0 }

(** Settles an arena read that cannot fail for a handle this module produced: every cached endpoint
    is a node this arena published, only labelled nodes are dereferenced, and every depth handed to
    the arena lies on the tail's own ancestry path. A failure here would mean the handle was
    corrupted, which the abstract type prevents, so it is reported as a defect rather than threaded
    through every result. *)
let settled what = function
  | Ok value -> value
  | Error message -> invalid_arg ("bilateral ancestral deque: " ^ what ^ ": " ^ message)

let value_of arena node = settled "value" (Incremental_ancestor.value_at arena node)
let depth_of arena node = settled "depth" (Incremental_ancestor.depth_of arena node)
let parent_of arena node = settled "parent" (Incremental_ancestor.parent_of arena node)

let ancestor_of arena node ~depth =
  settled "ancestor" (Incremental_ancestor.ancestor_at_depth arena node ~depth)

let create () =
  let empty = empty_at Incremental_ancestor.bottom in
  { arena = Incremental_ancestor.create (); left = empty; right = empty; total = 0 }

let of_arena arena =
  let* depth = Incremental_ancestor.depth_of arena Incremental_ancestor.bottom in
  if depth <> -1 then Error "the arena bottom node must have depth -1"
  else
    let empty = empty_at Incremental_ancestor.bottom in
    Ok { arena; left = empty; right = empty; total = 0 }

let arena deque = deque.arena
let count deque = deque.total
let is_empty deque = deque.total = 0

let shares_storage_with left right =
  left.arena == right.arena && left.left = right.left && left.right = right.right
  && left.total = right.total

(** An empty handle over the same arena, anchored at the arena's bottom node. *)
let empty_handle deque =
  let empty = empty_at Incremental_ancestor.bottom in
  { deque with left = empty; right = empty; total = 0 }

let first deque =
  if deque.total = 0 then None
  else
    let node = if deque.left.size <> 0 then deque.left.tail else deque.right.base in
    Some (value_of deque.arena node)

let last deque =
  if deque.total = 0 then None
  else
    let node = if deque.right.size <> 0 then deque.right.tail else deque.left.base in
    Some (value_of deque.arena node)

let nth index deque =
  if index < 0 || index >= deque.total then None
  else
    let node =
      if index < deque.left.size then
        if index = 0 then deque.left.tail
        else if index = deque.left.size - 1 then deque.left.base
        else
          ancestor_of deque.arena deque.left.tail
            ~depth:(depth_of deque.arena deque.left.tail - index)
      else
        let right_index = index - deque.left.size in
        if right_index = 0 then deque.right.base
        else if right_index = deque.right.size - 1 then deque.right.tail
        else
          ancestor_of deque.arena deque.right.tail
            ~depth:(depth_of deque.arena deque.right.anchor + right_index + 1)
    in
    Some (value_of deque.arena node)

(** Publishes one leaf below an interval's tail and extends the interval by one. *)
let add_to_tail arena segment value =
  let* tail = Incremental_ancestor.add_leaf arena ~parent:segment.tail value in
  let base = if segment.size = 0 then tail else segment.base in
  Ok { anchor = segment.anchor; base; tail; size = segment.size + 1 }

let add_first value deque =
  if deque.total = max_int then Error "the deque count cannot grow further"
  else
    let* left = add_to_tail deque.arena deque.left value in
    Ok { deque with left; total = deque.total + 1 }

let add_last value deque =
  if deque.total = max_int then Error "the deque count cannot grow further"
  else
    let* right = add_to_tail deque.arena deque.right value in
    Ok { deque with right; total = deque.total + 1 }

(** Drops an interval's newest physical label by moving its tail to the parent. No queries. *)
let remove_tail arena segment =
  let tail = parent_of arena segment.tail in
  if segment.size = 1 then empty_at tail
  else { anchor = segment.anchor; base = segment.base; tail; size = segment.size - 1 }

(** Drops an interval's oldest physical label by advancing its anchor to the cached base. Costs at
    most one level-ancestor query, which selects the next cached base; an interval of two values
    already holds that node as its tail. *)
let remove_base arena segment =
  if segment.size = 1 then empty_at segment.tail
  else
    let base =
      if segment.size = 2 then segment.tail
      else
        ancestor_of arena segment.tail ~depth:(depth_of arena segment.tail - segment.size + 2)
    in
    { anchor = segment.base; base; tail = segment.tail; size = segment.size - 1 }

let remove_first deque =
  if deque.total = 0 then None
  else if deque.total = 1 then Some (empty_handle deque)
  else if deque.left.size <> 0 then
    Some { deque with left = remove_tail deque.arena deque.left; total = deque.total - 1 }
  else Some { deque with right = remove_base deque.arena deque.right; total = deque.total - 1 }

let remove_last deque =
  if deque.total = 0 then None
  else if deque.total = 1 then Some (empty_handle deque)
  else if deque.right.size <> 0 then
    Some { deque with right = remove_tail deque.arena deque.right; total = deque.total - 1 }
  else Some { deque with left = remove_base deque.arena deque.left; total = deque.total - 1 }

let pop_first deque =
  match (first deque, remove_first deque) with
  | Some value, Some rest -> Some (value, rest)
  | _ -> None

let pop_last deque =
  match (last deque, remove_last deque) with
  | Some value, Some rest -> Some (value, rest)
  | _ -> None

(** Selects the sub-interval starting [start] values in and holding [size] values of a reversed left
    arm, where [start] counts down from the tail. At most two queries, and fewer whenever a cached
    endpoint already answers. *)
let slice_left arena segment start size =
  if size = 0 then empty_at Incremental_ancestor.bottom
  else if start = 0 && size = segment.size then segment
  else begin
    let tail_depth = depth_of arena segment.tail in
    let base_depth = tail_depth - segment.size + 1 in
    let sliced_tail_depth = tail_depth - start in
    let sliced_base_depth = sliced_tail_depth - size + 1 in
    let tail =
      if sliced_tail_depth = tail_depth then segment.tail
      else ancestor_of arena segment.tail ~depth:sliced_tail_depth
    in
    let base =
      if sliced_base_depth = sliced_tail_depth then tail
      else if sliced_base_depth = base_depth then segment.base
      else ancestor_of arena segment.tail ~depth:sliced_base_depth
    in
    let anchor = if base = segment.base then segment.anchor else parent_of arena base in
    { anchor; base; tail; size }
  end

(** Selects the sub-interval starting [start] values in and holding [size] values of a forward right
    arm. At most two queries, and fewer whenever a cached endpoint already answers. *)
let slice_right arena segment start size =
  if size = 0 then empty_at Incremental_ancestor.bottom
  else if start = 0 && size = segment.size then segment
  else begin
    let anchor_depth = depth_of arena segment.anchor in
    let sliced_base_depth = anchor_depth + start + 1 in
    let sliced_tail_depth = sliced_base_depth + size - 1 in
    let finish = start + size in
    let base =
      if start = 0 then segment.base
      else ancestor_of arena segment.tail ~depth:sliced_base_depth
    in
    let tail =
      if sliced_tail_depth = sliced_base_depth then base
      else if finish = segment.size then segment.tail
      else ancestor_of arena segment.tail ~depth:sliced_tail_depth
    in
    let anchor = if start = 0 then segment.anchor else parent_of arena base in
    { anchor; base; tail; size }
  end

let slice ~index ~count deque =
  if index < 0 then Error "the slice index is negative"
  else if count < 0 then Error "the slice count is negative"
  else if index > deque.total then Error "the slice index is outside the deque"
  else if index > deque.total - count then Error "the slice count is outside the deque"
  else if index = 0 && count = deque.total then Ok deque
  else if count = 0 then Ok (empty_handle deque)
  else begin
    let finish = index + count in
    let left_start = min index deque.left.size in
    let left_end = min finish deque.left.size in
    let right_start = max 0 (index - deque.left.size) in
    let right_end = max 0 (finish - deque.left.size) in
    let left = slice_left deque.arena deque.left left_start (left_end - left_start) in
    let right = slice_right deque.arena deque.right right_start (right_end - right_start) in
    Ok { deque with left; right; total = count }
  end

let take amount deque =
  if amount < 0 then Error "the prefix length is negative"
  else if amount > deque.total then Error "the prefix length is outside the deque"
  else if amount = deque.total then Ok deque
  else slice ~index:0 ~count:amount deque

let drop amount deque =
  if amount < 0 then Error "the omitted count is negative"
  else if amount > deque.total then Error "the omitted count is outside the deque"
  else if amount = 0 then Ok deque
  else slice ~index:amount ~count:(deque.total - amount) deque

let split_at index deque =
  if index < 0 then Error "the split boundary is negative"
  else if index > deque.total then Error "the split boundary is outside the deque"
  else
    let* prefix = take index deque in
    let* suffix = drop index deque in
    Ok (prefix, suffix)

let reverse deque =
  if deque.total <= 1 then deque else { deque with left = deque.right; right = deque.left }

let clear deque = if deque.total = 0 then deque else empty_handle deque

(** The values on the [length] nodes at and above [node], deepest ancestor first. The right arm
    reads in that order; the left arm reads in the reverse of it. *)
let rec ancestor_values arena node length collected =
  if length = 0 then collected
  else
    let collected = value_of arena node :: collected in
    if length = 1 then collected
    else ancestor_values arena (parent_of arena node) (length - 1) collected

let to_list deque =
  List.rev_append
    (ancestor_values deque.arena deque.left.tail deque.left.size [])
    (ancestor_values deque.arena deque.right.tail deque.right.size [])

let to_seq deque =
  let right_seq () =
    List.to_seq (ancestor_values deque.arena deque.right.tail deque.right.size []) ()
  in
  let rec left_seq node remaining () =
    if remaining = 0 then right_seq ()
    else
      let value = value_of deque.arena node in
      let next = if remaining = 1 then node else parent_of deque.arena node in
      Seq.Cons (value, left_seq next (remaining - 1))
  in
  left_seq deque.left.tail deque.left.size

let iter action deque =
  let node = ref deque.left.tail in
  for index = 0 to deque.left.size - 1 do
    action (value_of deque.arena !node);
    if index + 1 <> deque.left.size then node := parent_of deque.arena !node
  done;
  List.iter action (ancestor_values deque.arena deque.right.tail deque.right.size [])

let fold_left combine initial deque =
  let state = ref initial in
  let node = ref deque.left.tail in
  for index = 0 to deque.left.size - 1 do
    state := combine !state (value_of deque.arena !node);
    if index + 1 <> deque.left.size then node := parent_of deque.arena !node
  done;
  List.fold_left combine !state (ancestor_values deque.arena deque.right.tail deque.right.size [])

let of_list values =
  List.fold_left (fun deque value -> Result.bind deque (add_last value)) (Ok (create ())) values

(** Checks one interval's endpoint identities, cached base, and count/depth agreement. An empty
    interval is settled from cached fields alone; a nonempty one spends exactly two level-ancestor
    queries to place the anchor and the cached base on the tail's ancestry path. *)
let validate_segment arena segment =
  if segment.size = 0 then
    if segment.anchor <> segment.base || segment.base <> segment.tail then
      Error "an empty interval does not have one shared endpoint"
    else Ok ()
  else
    let* anchor_depth = Incremental_ancestor.depth_of arena segment.anchor in
    let* tail_depth = Incremental_ancestor.depth_of arena segment.tail in
    if tail_depth - anchor_depth <> segment.size then
      Error "an interval count disagrees with its endpoint depths"
    else
      let* base_parent = Incremental_ancestor.parent_of arena segment.base in
      if base_parent <> segment.anchor then
        Error "the cached base is not the first node after the anchor"
      else
        let* found_anchor =
          Incremental_ancestor.ancestor_at_depth arena segment.tail ~depth:anchor_depth
        in
        if found_anchor <> segment.anchor then
          Error "the interval anchor is not an ancestor of its tail"
        else
          let* found_base =
            Incremental_ancestor.ancestor_at_depth arena segment.tail ~depth:(anchor_depth + 1)
          in
          if found_base <> segment.base then
            Error "the cached base is not on the tail's ancestry path"
          else Ok ()

let validate deque =
  if deque.total <> deque.left.size + deque.right.size then
    Error "the total count disagrees with the interval counts"
  else
    let* () = validate_segment deque.arena deque.left in
    let* () = validate_segment deque.arena deque.right in
    Ok { count = deque.total; left_count = deque.left.size; right_count = deque.right.size }
